File: kanonak-codec/java/src/main/java/org/kanonak/codec/KanonakNode.java
package org.kanonak.codec;

import java.util.Map;

/**
 * The {@code $}-envelope as data — the base class a generated typed model
 * extends so an instance carries its own identity and binds straight to the
 * normalized-JSON wire form via {@link TypedNodes}. Envelope keys are reserved
 * (never ontology statements); {@link #getExtra() extra} holds the open-world
 * assertions outside the type-model, keyed by predicate URI, and rides the
 * wire as sibling fields (matching {@link Codec#serialize} semantics).
 */
public abstract class KanonakNode {
    private String id;
    private String typeUri;
    private String name;
    private String packageContentHash;
    private String packageVersion;
    private Map<String, Object> extra;

    /** The resource's canonical URI ({@code $id}). Required to form a subject. */
    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    /** The durable class URI ({@code $type}) — the value of the synthesized type triple. */
    public String getTypeUri() {
        return typeUri;
    }

    public void setTypeUri(String typeUri) {
        this.typeUri = typeUri;
    }

    /**
     * An embedded value's authored dict-key ({@code $name}) — HASH-RELEVANT
     * (serialized into the canonical form). Only meaningful when this instance
     * is used as an embedded value (via {@link Ref#embed(Object, String)});
     * {@code null} for subjects.
     */
    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    /** Package provenance on read ({@code $contentHash}); ignored if echoed back on write. */
    public String getPackageContentHash() {
        return packageContentHash;
    }

    public void setPackageContentHash(String packageContentHash) {
        this.packageContentHash = packageContentHash;
    }

    /** Package provenance on read ({@code $version}); ignored if echoed back on write. */
    public String getPackageVersion() {
        return packageVersion;
    }

    public void setPackageVersion(String packageVersion) {
        this.packageVersion = packageVersion;
    }

    /**
     * Open-world assertions outside the type-model, keyed by predicate URI.
     * Lossless round-trip: a consumer that ignores this still carries it.
     * May be {@code null} when the instance carries none.
     */
    public Map<String, Object> getExtra() {
        return extra;
    }

    public void setExtra(Map<String, Object> extra) {
        this.extra = extra;
    }
}

File: kanonak-codec/csharp/src/Kanonak.Codec/KanonakNode.cs
using System.Collections.Generic;
using System.Text.Json.Serialization;

namespace Kanonak.Codec
{
    /// <summary>
    /// The <c>$</c>-envelope as data — the base class a generated typed model
    /// extends so an instance carries its own identity and serializes straight
    /// to the normalized-JSON wire form via System.Text.Json. Envelope keys are
    /// reserved (never ontology statements); <see cref="Extra"/> holds the
    /// open-world assertions outside the type-model, keyed by predicate URI,
    /// and rides the wire as sibling fields (<c>[JsonExtensionData]</c>
    /// semantics — matching <see cref="Codec.Serialize"/>).
    /// </summary>
    public abstract class KanonakNode
    {
        /// <summary>The resource's canonical URI. Required to form a subject.</summary>
        [JsonPropertyName("$id")]
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
        public string Id { get; set; }

        /// <summary>The durable class URI — the value of the synthesized type triple.</summary>
        [JsonPropertyName("$type")]
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
        public string Type { get; set; }

        /// <summary>
        /// A multi-typed node's FULL type set (0.4.0, runtime#10) — present only
        /// when the node carries more than one type statement. Sorted by UTF-8
        /// bytes, at least two members, no duplicates, <see cref="Type"/> a
        /// member; each member emits one type statement in canonical form.
        /// Exposed ONLY as the <c>$types</c> envelope — deliberately no
        /// unprefixed wire name, because an ontology can model a property
        /// literally named <c>types</c>; the <c>$</c> prefix exists to avoid
        /// exactly that collision.
        /// </summary>
        [JsonPropertyName("$types")]
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
        public List<string> Types { get; set; }

        /// <summary>
        /// An embedded value's authored dict-key — HASH-RELEVANT (serialized into
        /// the canonical form). Only meaningful when this instance is used as an
        /// embedded value (via <see cref="Ref{T}.Embed(T)"/>); null for subjects.
        /// </summary>
        [JsonPropertyName("$name")]
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
        public string Name { get; set; }

        /// <summary>Package provenance on read; ignored if echoed back on write.</summary>
        [JsonPropertyName("$contentHash")]
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
        public string PackageContentHash { get; set; }

        /// <summary>Package provenance on read; ignored if echoed back on write.</summary>
        [JsonPropertyName("$version")]
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
        public string PackageVersion { get; set; }

        /// <summary>
        /// Open-world assertions outside the type-model, keyed by predicate URI.
        /// Lossless round-trip: a consumer that ignores this still carries it.
        /// </summary>
        [JsonExtensionData]
        public Dictionary<string, object> Extra { get; set; }
    }
}
